README.md
# mailChain
Mailchain is a Python Script that create a SMTP-Relay Server and decides based on some Rules from a MYSQL Database which whey the E-Mail go (another SMTP Server or a HTTP Call).

It creates a new MYSQL-Connection (and Query) for each Mail, so it always have the newest Version but it take some time (and load)

## Functions
### Basic Functions
* Open SMTP-Relay Server
* Check Mail agains Rule to decide which SMTP-Relay or HTTP Call used to forward to
* Check E-Mails bevor Forwording agains spamc
* Log Mails to detect spam
* Log complete Mail inc. Body to mysql

### Posible Rules in on Chain
* Check if From or Subject Match a RegEx (From MYSQL)
* Check if one or all to Header Match a RegEx (From MYSQL)

### Extra Rules
* Check if AuthenticatedSende in Mail Header Match a Regex (From MYSQL)

### Modify E-Mail
* Remove AuthenticatedSende from Header bevor forward it
* Add a Received Header to the Mail (Include a Name and the Chain-ID)

### Forward a Mail
* (Re)send Abuse Message to Sender (if AuthenticatedSende is required and wrong)
* Send Mail to another SMTP-Relay
* Make a HTTP Call with the Mail Content in a POST Request
* Log Mail (To, From, Subject, AuthenticatedSende, Timestamp) to a MYSQL Database to detect Spaming
* Send Mails to the receiving mailserver (testing)

## Config File
The Basic Config File contains the MYSQL Connection and some Basic Config. Check the defaults.cfg.sample for more detais.

### MYSQL
*  host -> Contains the MYSQL Host
*  user -> Contains the MYSQL User
*  pass -> Contains the MYSQL Password
*  db -> Contains the MYSQL Database Name

### Mail
* dump -> Its old and not used any more, ignore it
* addReceived -> If it "True" a "Received" Entry to the Mail Header
* ReceivedName -> Contains the Name to Add in the "Received" Entry
* sendLog -> If it is True, all outgoing Mails log in a MYSQL Database
* mailAuthenticatedSender -> If it is "True" outgoing Mails Check if a AuthenticatedSender is required and if it match
* spamc -> Check if the mail with spamc and reject if the Spam-Sore is to hight (for all mails)
* dumpSQL -> Dump all mails in the MYSQL-Database

### SendAbuse
If the Script have to send a abused Message (e.g. if the AuthenticatedSender don't match) it use the SMTP-Setting of this Section to send it.

### sendMail
This is a testing feature to send the mail not to another relay, send it direct to the receiving Mailserver.

It only works if mailLog and dumpSQL are enabled!

* enabled -> Enabled or Disabled this feature
* loopTime -> how much minuts should be wait bevor next check of the DB for new E-Mails

## Installation
Do the following Steps:

###Install Requirments
```
sudo apt-get install python3.4-venv libmysqlclient-dev python3-dev
pyvenv venv
source venv/bin/acticate
pip install -r requirments.txt
```
### Set up the Database
Create the Database Structure from the mysql.sql File and create a new MYSQL User

### Clone the Project and Config it
Clone the Project and change the defaults.cfg to your Settings. Than run the server, install all needet dependencies. Than Start the server.py

### Config Postfiy
Let Postfiy relay all Mails ofter 127.0.0.1:4467

### Add a Default Rule in your Chain
```
INSERT INTO `mailChain` (prio`, `from`, `to`, `allTo`, `subject`, `smtpServer`, `smtpPort`, `smtpUser`, `smtpPass`, `httpCall`, `removeAuthenticatedSender`) VALUES
(100000, NULL, NULL, 1, NULL, 'smtp-relay.server', NULL, NULL, NULL, NULL, 1);
```


Changelog.md
#dev-threading
* rewrite Rules
* switch to pymysql (python 3)
* switch to configparser (python3)
* Performanter Config Object (read single time)
* Performanter MYSQL Connection (one connection possible)
* OOP
* add Debug Modus (dont send abuse mails)
* add dumpToMYSQL
* remove hdd dump
* send without relay (beta)
* mailLog is can not be disabled anymore

#dev-master
* add spamC parameter (general)

#0.1.2
* Update to Python 3
* use pip to install inbox.py
* use sender to send abuse mails
* add General "sendMail" function
* exception handling with abuse mails
* Return error codes


run.py
#!/usr/bin/python3
from threadController.relayController import relayController
from threadController.sendMail import sendMail
import logging
import time
import configparser
import os
import pymysql


logger = logging.getLogger()
handler = logging.StreamHandler()
formatter = logging.Formatter('%(asctime) 12s %(levelname)-8s %(message)s')
handler.setFormatter(formatter)
logger.addHandler(handler)
logger.setLevel(logging.DEBUG)

logger.info("Start")
logger.debug("Try to find "+os.path.dirname(os.path.realpath(__file__))+"/config/defaults.cfg")
if os.path.isfile(os.path.dirname(os.path.realpath(__file__))+"/config/defaults.cfg") == False:
	logger.error("Config file not found")
	raise Exception("Config not found")

config = configparser.RawConfigParser()
config.read(os.path.dirname(os.path.realpath(__file__))+'/config/defaults.cfg')

logger.debug("Check MYSQL Connection")
db = pymysql.connect(host=config.get('MYSQL', 'host'), user=config.get('MYSQL', 'user'), passwd=config.get('MYSQL', 'pass'), db=config.get('MYSQL', 'db'))
t = relayController()
t.start()
if(config.get('sendMail', 'enabled')=="True"):
	m = sendMail()
	m.start()
	m.join()
t.join()
print("Main script Stops!")


server.py
#!/usr/bin/python
from inbox import Inbox
import MySQLdb as mdb
from smtplib import SMTP
import requests
from backports import configparser as ConfigParser
import os
#import ConfigParser, os
import re
import datetime
import sys
import hashlib
from sender import Mail, Message
import time
import tempfile
import uuid

inbox = Inbox()

def checkChain(rule, to, sender, subject):
    check = True #Status of this Rule
    #Check if Sender Match
    if rule[2] != None and re.match(rule[2], sender) is None:
        check = False
    #Check if To Match
    oneToFails = False #If it get True, min on to header don`t match
    oneToMatch = False #If it get True, min on to header match
    for mail in to:
        if rule[3] != None and re.match(rule[3], mail) is None:
            oneToFails = True
        elif rule[3] != None and re.match(rule[3], mail) is not None:
            oneToMatch = True
    if rule[4] == True and oneToFails == True: #Rules says all to header must match but they dont do
        check = False
    elif rule[4] == False and oneToMatch == False: #Rules say min one to header must match but no to header matchs
        check = False
    #Check Subject
    if rule[5] != None and re.match(rule[5], subject) is None:
        check = False
    return check

def checkSpamc(sender, body):
    filename = tempfile.gettempdir()+"/"+str(uuid.uuid1())
    with open(filename, 'w') as f:
        f.write(body)
    ret = os.popen("spamc -c < %s" % filename).read()
    detais = str(ret).split("/")
    count = float(detais[0].strip())
    max = float(detais[1].strip())
    pLog("SpamC: "+ret)
    if count > max:
        return False
    return True

def checkAuthenticatedSender(sender, body, asenderRegEx):
    pLog("Check Authenticated Sender")
    #print(asenderRegEx)
    asender = re.findall(r'\(Authenticated\ssender\:\s{0,10}([^)]*)\)\n\s*', body)
    if(len(asender)>0):
        asender = asender[0]
    else:
        asender = None
    if asenderRegEx != None and re.match(asenderRegEx, str(asender)) is None:
        pLog("AuthenticatedSender is required but failed")
        #Send Abuse Mail
        config = ConfigParser.RawConfigParser()
        config.read('defaults.cfg')
        msg = "Mail from %s send from Authenticated Sender %s\r\n\r\nThe Mail Header must match!\r\nMail don't send, please check the Config or Contact your Mail-Server-Admin per E-Mail to %s" % (sender, asender, config.get('SendAbuse', 'from'))
        sendMail(sender, "Mail from "+sender+" with wrong Authenticated Sender Header", msg)
        pLog("AuthenticatedSender %s dont match on %s" % (asender, asenderRegEx))
        return False
    pLog("AuthenticatedSender Match")
    return True

def dumpMail(to, sender, subject, body):
    if not os.path.exists("./mails"):
        os.makedirs("./mails")
    if not os.path.exists("./mails/%s" % sender):
        os.makedirs("./mails/%s" % sender)
    d = datetime.datetime.now()
    m = hashlib.md5()
    m.update("%s_%s" % (d.strftime("%y_%m_%d_%H_%M_%S"), subject))
    fp = open("./mails/%s/%s.txt" % (sender, str(m.hexdigest())), "w")
    fp.write(body)
    fp.close()
    pLog("Dump Mail to File System")

def logMail(to, sender, subject, body, con):
    asender = re.findall(r'\(Authenticated\ssender\:\s{0,10}([^)]*)\)\n\s*', body)
    cur = con.cursor()
    if(len(asender)>0):
        asender = asender[0]
    else:
        asender = None
    for t in to:
        sql = "INSERT INTO `mailLog`(`from`, `to`, `authenticatedSender`, `subject`) VALUES ('%s', '%s', '%s', '%s')" % (sender, t, asender, subject)
        cur.execute(sql)
    con.commit()
    pLog("Mail Logged in DB")

def removeAuthenticatedSende(body):
    body = re.sub(r'\(Authenticated\ssender\:\s[^)]*\)\n\s*', '', body)
    pLog("Remove Authenticated Sende from Mail Head")
    return body

def pLog(msg):
    d = datetime.datetime.now()
    print("[%s] %s" % (d.strftime("%y-%m-%d %H:%M:%S"), msg))

def addReceived(body, ruleId):
    config = ConfigParser.RawConfigParser()
    config.read('defaults.cfg')
    start = body.find("Received")
    receivedmsg = "Received: mailChain ("+config.get('Mail', 'ReceivedName')+") #"+str(ruleId)+"\r\n"
    body = body[0:start] + receivedmsg + body[start:]
    pLog("Add Received Header to Mail")
    return body

def runBasic(to, sender, subject, body):
    pLog("Incoming mail from %s" % sender)
    #Read Config
    config = ConfigParser.RawConfigParser()
    config.read('defaults.cfg')
    #Dump Mail to File System
    if config.get('Mail', 'dump') == True:
        dumpMail(to, sender, subject, body)
    #Connect to Database
    try:
        con = mdb.connect(config.get('MYSQL', 'host'), config.get('MYSQL', 'user'), config.get('MYSQL', 'pass'), config.get('MYSQL', 'db'));
        cur = con.cursor()
    except:
        pLog("FATAL!!! MYSQL Connection cant create, can`t run without MYSQL")
        pLog(str(sys.exc_info()))
        sys.exit()
    #Get Rules
    cur.execute("SELECT * FROM `mailChain` ORDER BY `mailChain`.`prio` ASC")
    chains = cur.fetchall()
    send = False
    for rule in chains:
        if send == True:
            continue
        check = checkChain(rule, to, sender, subject)
        pLog("Chain %s is %s" % ((rule[0]), (check)))
        if check == True:
            #Send Mail
            send = True     #Mail alwady send
            tosend = True   #Send Mail?

            #Do Stuff with Mail
            #Add Received Header to Mail
            if config.get('Mail', 'addReceived'):
                body = addReceived(body, rule[0])
            #Check Authenticated Header
            if config.get('Mail', 'mailAuthenticatedSender'):
                pLog("Check Authentication Sender for %s" % sender)
                cur.execute('SELECT * FROM `mailAuthenticatedSender` WHERE `from` = "%s"' % sender)
                asenderregex = cur.fetchone()
                if asenderregex != None:
                    retAuth = checkAuthenticatedSender(sender, body, str(asenderregex[2]))
                    if retAuth == False:
                        tosend = False
                        return "503 Autoriced Header is wrong"
            if config.get('Mail', 'spamc') == True:
                res = checkSpamc(sender, body)
                if res == False:
                    pLog("Mail as Spam detectet and reject")
                    tosend = False
                    return "550 Spam detect"
                pLog("No Spam, send mail")
            tosend = False
            if tosend == True:
                pLog("Redirect Mail")
                if config.get('Mail', 'sendLog'):
                    logMail(to, sender, subject, body, con)
                #Remove Authenticate Header:
                if rule[11] == True:
                    body = removeAuthenticatedSende(body)
                if rule[6] != None:
                    pLog("Send Mail ofer SMTP Relay %s" % rule[6])
                    smtp = SMTP()
                    port = 25
                    if rule[7] != None:
                        port = int(rule[7])
                    smtp.connect(str(rule[6]), int(port))
                    if rule[8] != None and rule[9] != None:
                        smtp.login(rule[8], rule[9])
                    smtp.sendmail(sender, to, body)
                    smtp.quit()
                    return "250 OK"
                if rule[10] != None:
                    pLog("Make HTTP Call to %s" % rule[10])
                    payload = {'to': to, 'sender': sender, 'subject': subject, 'body': body}
                    r = requests.post(rule[10], data=payload)
                    return "250 OK - POST Request send"
            else:
                return "550 Something go wrong"

def sendMail(to, subject, msgContext):
    pLog("Send Server Mail")
    config = ConfigParser.RawConfigParser()
    config.read('defaults.cfg')
    msg = Message(subject, fromaddr=config.get('SendAbuse', 'from'), to=to)
    if(to!=config.get('SendAbuse', 'from')):
        msg.bcc = config.get('SendAbuse', 'from')
    msg.body = msgContext
    msg.date = time.time()
    msg.charset = "utf-8"
    mail = Mail(config.get('SendAbuse', 'server'), port=config.get('SendAbuse', 'port'), username=config.get('SendAbuse', 'user'), password=config.get('SendAbuse', 'pass'),use_tls=False, use_ssl=False, debug_level=None)
    mail.send(msg)

@inbox.collate
def handle(to, sender, subject, body):
    try:
        return runBasic(to, sender, subject, body)
    except:
        pLog("Something go Wrong, Mail sendet")
        errorStr = str(sys.exc_info())
        config = ConfigParser.RawConfigParser()
        config.read('defaults.cfg')
        sendMail(sender, "Mail cant not sendet", "Hello, you try to send a Mail, we cant send this Mail because an error appear. Our Team get the Information, please try it again later!")
        sendMail(config.get('SendAbuse', 'from'), "MailChain Error", "So following error appear:\r\n\r\n"+errorStr)
        return "550 Something go wrong"

# Bind directly.
pLog("Start Server")
#sendMail("hello@kekskurse.de", "Abuse Mail", "Something go Wrong!")
inbox.serve(address='127.0.0.1', port=4467)


threadController/sendMail.py
import threading
import time
import logging
import pymysql
import configparser
import dns.resolver
import smtplib
import sys
import tempfile
import uuid
import urllib.request
import traceback
from lib.mailChainSMTP import MailChainSMTP

class sendMail(threading.Thread):
	def __init__(self, config = None, db = None):
		self.log = logging.getLogger()
		self.log.debug("Init sendMail Controll Thread")
		threading.Thread.__init__(self)
		if(config == None):
			self.config = configparser.RawConfigParser()
			self.config.read('config/defaults.cfg')
		else:
			self.config = config
		if(db == None):
			try:
				self.db = pymysql.connect(host=self.config.get('MYSQL', 'host'), user=self.config.get('MYSQL', 'user'), passwd=self.config.get('MYSQL', 'pass'), db=self.config.get('MYSQL', 'db'))
			except:
				self.log.critical("MYSQL Connection can't be estaplet")
				self.setResponse("550 - Server Error")
		else:
			self.db = db
		self.cur = self.db.cursor()

	def sendMail(self, to, subject, msgContext):
		self.log.debug("Send Server Mail")
		if(self.config.get('SendAbuse', 'debug') == "True"):
			self.log.info("Debug modus, dont send Mails!")
			return False
		msg = Message(subject, fromaddr=self.config.get('SendAbuse', 'from'), to=to)
		if(to!=self.config.get('SendAbuse', 'from')):
			msg.bcc = self.config.get('SendAbuse', 'from')
		else:
			mailattachment = Attachment("orginalmail.txt", "text/plain", self.body)
			msg.attach(mailattachment)
		msg.body = msgContext
		msg.date = time.time()
		msg.charset = "utf-8"
		mail = Mail(self.config.get('SendAbuse', 'server'), port=self.config.get('SendAbuse', 'port'), username=self.config.get('SendAbuse', 'user'), password=self.config.get('SendAbuse', 'pass'),use_tls=False, use_ssl=False, debug_level=None)
		mail.send(msg)
		self.log.debug("Mail to %s sended" % to)
	def run(self):
		while(True):
			self.log.debug("SendMails")
			self.cur.execute("SELECT * FROM `outgoing` WHERE `status` = 'toSent'")
			res = self.cur.fetchall()
			for mail in res:
				try:
					self.log.debug("Send Mail #"+str(mail[0]))
					domain = mail[1][mail[1].find("@")+1:]
					self.log.debug("Domain: "+domain)
					mailserver = dns.resolver.query(domain, 'MX')
					self.log.debug("Mailserver: "+str(mailserver[0].exchange))
					smtpObj = MailChainSMTP(str(mailserver[0].exchange))
					smtpObj.set_debuglevel(1)
					try:
						res = smtpObj.sendmail(mail[2], mail[1], mail[4])
						self.cur.execute("UPDATE `mail2`.`outgoing` SET `status` = 'sent' WHERE `outgoing`.`id` = "+str(int(mail[0]))+";")
					except:
						errorStr = traceback.format_exc()
						self.sendMail(mail[2], "Mail cant send", "Sorry, this is a beta function and it dont work!!\r\n\r\n"+errorStr)
						self.cur.execute("UPDATE `mail2`.`outgoing` SET `status` = 'error' WHERE `outgoing`.`id` = "+str(int(mail[0]))+";")
						self.cur.execute("INSERT INTO `sendMailLog`(`mailID`, `log`) VALUES (%s, %s)", (int(mail[0]), errorStr))
						self.log.info("Mail can not send")
					finally:
						log = smtpObj.getLog();
						logStr = ""
						for l in log:
							logStr += l+"\r\n"
						self.cur.execute("INSERT INTO `sendMailLog`(`mailID`, `log`) VALUES (%s, %s)", (int(mail[0]), logStr))
						self.log.debug("Mail Send")

				except:
					errorStr = traceback.format_exc()
					self.log.error("Mail Send failed!")
					self.log.error(errorStr)
					self.cur.execute("INSERT INTO `sendMailLog`(`mailID`, `log`) VALUES (%s, %s)", (int(mail[0]), errorStr))
					self.cur.execute("UPDATE `mail2`.`outgoing` SET `status` = 'error' WHERE `outgoing`.`id` = "+str(int(mail[0]))+";")
					self.log.debug("Mail Send")

			self.db.commit()
			time.sleep(int(self.config.get('sendMail', 'loopTime')))


threadController/relayController.py
import threading
import logging
from inbox import Inbox
import time
import pymysql
import configparser
from threadController.chainController import chainController
import sys

inbox = Inbox()

class relayController(threading.Thread):
	def __init__(self):
		self.log = logging.getLogger()
		self.log.debug("Init relay Controll Thread")
		threading.Thread.__init__(self)
		self.config = configparser.RawConfigParser()
		self.config.read('config/defaults.cfg')

	def run(self):
		self.log.debug("Run rely Controll Thread")
		inbox.serve(address='127.0.0.1', port=4467)

	@inbox.collate
	def handle(to, sender, subject, body):
		try:
			log = logging.getLogger()
			log.info("Incomming Mail")
			# Seperate in secound thread, than i can replace it faster
			t = chainController(to, sender, subject, body)
			t.start()
			t.join()
			response = t.retMailInfo()
			if(response == None):
				response = "550 - Oups"
			return response
		except:
			log.critical("Something go terrible Wrong: "+str(sys.exc_info()))
			return "550 Server Error"


threadController/chainController.py
import threading
import logging
import configparser
import pymysql
import sys
#from smtplib import SMTP
from sender import Mail, Message, Attachment
import requests
import re
import time
import traceback
import tempfile
import uuid
import os
from lib.mailChainSMTP import MailChainSMTP

class chainController(threading.Thread):
	def __init__(self, to, sender, subject, body, config = None, db = None):
		threading.Thread.__init__(self)
		self.to = to
		self.sender = sender
		self.subject = subject
		self.body = body
		self.messageID = None
		self.retStr = None
		self.log = logging.getLogger()
		self.log.debug("Init chain Controll Thread")
		self.mailLogId = None
		if(config == None):
			self.config = configparser.RawConfigParser()
			self.config.read('config/defaults.cfg')
		else:
			self.config = config
		if(db == None):
			try:
				self.db = pymysql.connect(host=self.config.get('MYSQL', 'host'), user=self.config.get('MYSQL', 'user'), passwd=self.config.get('MYSQL', 'pass'), db=self.config.get('MYSQL', 'db'))
			except:
				self.log.critical("MYSQL Connection can't be estaplet")
				self.setResponse("550 - Server Error")
		else:
			self.db = db
		self.cur = self.db.cursor()

	def sendMail(self, to, subject, msgContext):
		self.log.debug("Send Server Mail")
		if(self.config.get('SendAbuse', 'debug') == "True"):
			self.log.info("Debug modus, dont send Mails!")
			return False
		msg = Message(subject, fromaddr=self.config.get('SendAbuse', 'from'), to=to)
		if(to!=self.config.get('SendAbuse', 'from')):
			msg.bcc = self.config.get('SendAbuse', 'from')
		else:
			mailattachment = Attachment("orginalmail.txt", "text/plain", self.body)
			msg.attach(mailattachment)
		msg.body = msgContext
		msg.date = time.time()
		msg.charset = "utf-8"
		mail = Mail(self.config.get('SendAbuse', 'server'), port=self.config.get('SendAbuse', 'port'), username=self.config.get('SendAbuse', 'user'), password=self.config.get('SendAbuse', 'pass'),use_tls=False, use_ssl=False, debug_level=None)
		mail.send(msg)
		self.log.debug("Mail to %s sended" % to)

	def dumpMailHDD(self):
		self.log.error("dumpMail/dumpMailHDD is remved!")
		return False

	def dumpMailSQL(self):
		self.log.debug("Dump to DB")
		sql = 'INSERT INTO `outgoing`(`to`, `sender`,`orginalbody`, `body`, `status`) VALUES (%s, %s, %s, %s, "incommed")';
		self.cur.execute(sql, (self.to, self.sender, self.body, self.body))
		self.cur.execute("SELECT LAST_INSERT_ID()")
		id = self.cur.fetchone()
		self.log.info("Mail ID:"+str(id[0]))
		self.messageID = id[0]

	def retMailInfo(self):
		return self.retStr

	def addReceived(self, ruleId):
		start = self.body.find("Received")
		receivedmsg = "Received: mailChain ("+self.config.get('Mail', 'ReceivedName')+") #"+str(ruleId)+"\r\n"
		self.body = self.body[0:start] + receivedmsg + self.body[start:]
		self.log.info("Add Received Header to Mail")

	def checkAuthenticatedSender(self, asenderRegEx):
	    self.log.debug("Check Authenticated Sender")
	    asender = re.findall(r'\(Authenticated\ssender\:\s{0,10}([^)]*)\)\n\s*', self.body)
	    if(len(asender)>0):
	        asender = asender[0]
	    else:
	        asender = None
	    if asenderRegEx != None and re.match(asenderRegEx, str(asender)) is None:
	        self.log.info("AuthenticatedSender is required but failed")
	        #Send Abuse Mail
	        msg = "Mail from %s send from Authenticated Sender %s\r\n\r\nThe Mail Header must match!\r\nMail don't send, please check the Config or Contact your Mail-Server-Admin per E-Mail to %s" % (self.sender, asender, self.config.get('SendAbuse', 'from'))
	        self.sendMail(self.sender, "Mail from "+self.sender+" with wrong Authenticated Sender Header", msg)
	        self.log.debug("AuthenticatedSender %s dont match on %s" % (asender, asenderRegEx))
	        return False
	    self.log.debug("AuthenticatedSender Match")
	    return True

	def getAuthenticatedSenderCheck(self):
		self.log.debug("Get AuthenticatedSender Database")
		self.cur.execute('SELECT * FROM `mailAuthenticatedSender` WHERE `from` = "%s"' % self.sender)
		asenderregex = self.cur.fetchone()
		if asenderregex != None:
			retAuth = self.checkAuthenticatedSender(str(asenderregex[2]))
			if retAuth == False:
				self.setStatus("error")
				self.setResponse("503 Autoriced Header is wrong")
				return False
			else:
				return True

	def checkChain(self, rule):
		self.log.debug("Check Rule "+str(rule[0]))
		#Check if Sender Match
		if rule[2] != None and re.match(rule[2], self.sender) is None:
			return False
		#Check if To Match
		oneToFails = False #If it get True, min on to header don`t match
		oneToMatch = False #If it get True, min on to header match
		for mail in self.to:
			if rule[3] != None and re.match(rule[3], mail) is None:
				oneToFails = True
			elif rule[3] != None and re.match(rule[3], mail) is not None:
				oneToMatch = True
		if rule[4] == True and oneToFails == True: #Rules says all to header must match but they dont do
			return False
		elif rule[4] == False and oneToMatch == False: #Rules say min one to header must match but no to header matchs
			return False
		#Check Subject
		if rule[5] != None and re.match(rule[5], self.subject) is None:
			return False
		self.log.debug("Rule Match")
		return True

	def checkSpamc(self):
		self.log.debug("Check Spam")
		filename = tempfile.gettempdir()+"/"+str(uuid.uuid1())
		with open(filename, 'w') as f:
			f.write(self.body)
		ret = os.popen("spamc -c < %s" % filename).read()
		detais = str(ret).split("/")
		count = float(detais[0].strip())
		max = float(detais[1].strip())
		self.log.debug("Spam Score: "+ret.strip())
		if count > max:
			self.setStatus("error")
			self.setResponse("550 Spam detect")
			self.log.info("Spam Detact, reject")
			return False
		return True

	def removeAuthenticatedSende(self):
		self.body = re.sub(r'\(Authenticated\ssender\:\s[^)]*\)\n\s*', '', self.body)
		self.log.debug("Remove Authenticated Sende from Mail Head")

	def logMail(self):
		asender = re.findall(r'\(Authenticated\ssender\:\s{0,10}([^)]*)\)\n\s*', self.body)
		#cur = con.cursor()
		if(len(asender)>0):
			asender = asender[0]
		else:
			asender = None
		for t in self.to:
			#sql = "INSERT INTO `mailLog`(`from`, `to`, `authenticatedSender`, `subject`) VALUES ('%s', '%s', '%s', '%s')" % (self.sender, self.t, asender, self.subject)
			sql = "INSERT INTO `mailLog`(`from`, `to`, `authenticatedSender`, `subject`, `outgoingID`) VALUES (%s, %s, %s, %s, %s)"
			self.cur.execute(sql, (self.sender, t, asender, self.subject, self.messageID))
			if self.mailLogId == None:
				self.cur.execute("SELECT LAST_INSERT_ID()")
				id = self.cur.fetchone()
				self.mailLogId = id[0]
		self.db.commit()
		self.log.debug("Mail Logged in DB")

	def setStatus(self, status):
		if(self.config.get("Mail", "dumpSQL") == "True"):
			self.cur.execute("UPDATE `mail2`.`outgoing` SET `status` = %s WHERE `outgoing`.`id` = "+str(int(self.messageID))+";", (status))
			self.log.debug("Set Status to: "+str(status))
		else:
			self.log.error("Cant change status if dumpSQL is disabled")


	def runMain(self):
		# Dump Mail to HDD
		if self.config.get('Mail', 'dump') == "True":
			self.dumpMailHDD()
		if self.config.get("Mail", "dumpSQL") == "True":
			self.dumpMailSQL()
		# Check Auterised Sender Header
		if self.config.get('Mail', 'mailAuthenticatedSender'):
			ret = self.getAuthenticatedSenderCheck()
			if(ret == False):
				return False
		# General SpamC Check
		#print(self.config.get('Mail', 'spamc'))
		if self.config.get('Mail', 'spamc') == True or self.config.get('Mail', 'spamc') == "True":
			res = self.checkSpamc()
			if res == False:
				return False
		# Get Rules from DB
		self.cur.execute("SELECT * FROM `mailChain` ORDER BY `mailChain`.`prio` ASC")
		chains = self.cur.fetchall()
		for rule in chains:
			check = self.checkChain(rule)
			if check == True:
				# Add Recevied Header to Mail
				if self.config.get('Mail', 'addReceived'):
					self.addReceived(rule[0])
				self.logMail()
				if rule[11] == True:
					self.removeAuthenticatedSende()
				if rule[6] != None:
					self.log.info("Send via SMTP")
					smtp = MailChainSMTP()
					port = 25
					if rule[7] != None:
						port = int(rule[7])
					self.log.debug("SMTP-Server: "+str(rule[6])+":"+str(port))
					smtp.set_debuglevel(1)
					smtp.connect(str(rule[6]), int(port))

					if rule[8] != None and rule[9] != None:
						smtp.login(rule[8], rule[9])
					smtp.sendmail(self.sender, self.to, self.body)
					smtp.quit()
					log = smtp.getLog();
					logStr = ""
					for l in log:
						logStr += l+"\r\n"
					self.cur.execute("INSERT INTO `sendMailLog`(`mailID`, `log`) VALUES (%s, %s)", (int(self.mailLogId), logStr))
					self.log.debug("Mail Send")
					self.setResponse("250 OK")
					self.setStatus("delivered")
				if rule[10] != None:
					self.log.info("Make HTTP Call to %s" % rule[10])
					payload = {'to': to, 'sender': sender, 'subject': subject, 'body': body}
					r = requests.post(rule[10], data=payload)
					self.setResponse("250 OK - POST Request")
					self.setStatus("delivered")
				if rule[12] == True and rule[6] == None:
					if(self.config.get('sendMail', 'enabled')!="True"):
						self.setStatus("error")
						self.setResponse("550 - sendMail function not enabled")
					elif self.config.get("Mail", "dumpSQL") != "True":
						self.setStatus("error")
						self.setResponse("550 - enabled dumpSQL for direct Send")
					else:
						self.setStatus("toSent")
						self.setResponse("250 - sending")



				return True

	def setResponse(self, msg):
		if(self.retStr == None):
			self.retStr = str(msg)

	def run(self):
		self.log.debug("Run Mail Chain")
		try:
			self.runMain()
		except:
			errorStr = traceback.format_exc()
			self.log.critical(errorStr)
			self.sendMail(self.sender, "Mail cant not sendet", "Hello, you try to send a Mail, we cant send this Mail because an error appear. Our Team get the Information, please try it again later!")
			self.sendMail(self.config.get('SendAbuse', 'from'), "MailChain Error", "So following error appear:\r\n\r\n"+errorStr)
			self.setStatus("error")
			self.setResponse("550 - Oups")
		self.log.debug("Status: "+str(self.retStr))
		self.db.commit()
		#Mail Dump FileSystem
